File: app/csharp/tests/Ch09/StreamProcessingTests.cs
using Ch09;
using LanguageExt;
using static LanguageExt.Prelude;
using Xunit;

namespace GrokkingFP.CSharp.Tests.Ch09;

/// <summary>
/// 第9章: ストリーム処理のテスト
/// </summary>
public class StreamProcessingTests
{
    // ============================================
    // 基本的な Seq のテスト
    // ============================================

    [Fact]
    public void Numbers_Returns123()
    {
        var result = StreamProcessing.Numbers.ToList();
        Assert.Equal(new[] { 1, 2, 3 }, result);
    }

    [Fact]
    public void OddNumbers_FiltersEvenNumbers()
    {
        var input = Seq(1, 2, 3, 4, 5);
        var result = StreamProcessing.OddNumbers(input).ToList();
        Assert.Equal(new[] { 1, 3, 5 }, result);
    }

    [Fact]
    public void Doubled_DoublesEachNumber()
    {
        var input = Seq(1, 2, 3);
        var result = StreamProcessing.Doubled(input).ToList();
        Assert.Equal(new[] { 2, 4, 6 }, result);
    }

    // ============================================
    // 無限シーケンスのテスト
    // ============================================

    [Fact]
    public void Repeat_RepeatsInfinitely()
    {
        var result = StreamProcessing.Repeat(new[] { 1, 2, 3 }).Take(8).ToList();
        Assert.Equal(new[] { 1, 2, 3, 1, 2, 3, 1, 2 }, result);
    }

    [Fact]
    public void Infinite123s_RepeatsInfinitely()
    {
        var result = StreamProcessing.Infinite123s.Take(7).ToList();
        Assert.Equal(new[] { 1, 2, 3, 1, 2, 3, 1 }, result);
    }

    [Fact]
    public void Naturals_StartsFrom1()
    {
        var result = StreamProcessing.Naturals.Take(5).ToList();
        Assert.Equal(new[] { 1, 2, 3, 4, 5 }, result);
    }

    [Fact]
    public void From_StartsFromGivenValue()
    {
        var result = StreamProcessing.From(10).Take(5).ToList();
        Assert.Equal(new[] { 10, 11, 12, 13, 14 }, result);
    }

    // ============================================
    // サイコロストリームのテスト
    // ============================================

    [Fact]
    public void InfiniteDieCasts_ReturnsBetween1And6()
    {
        var results = StreamProcessing.InfiniteDieCasts.Take(10).ToList();
        Assert.True(results.All(n => n >= 1 && n <= 6));
    }

    [Fact]
    public void CastUntil_EndsWithTarget()
    {
        var results = StreamProcessing.CastUntil(6).Take(20).ToList();
        Assert.True(results.Count >= 1 && results.Count <= 20);
    }

    // ============================================
    // ストリーム操作のテスト
    // ============================================

    [Fact]
    public void Take_TakesFirstN()
    {
        var input = new[] { 1, 2, 3, 4, 5 };
        var result = StreamProcessing.Take(3, input).ToList();
        Assert.Equal(new[] { 1, 2, 3 }, result);
    }

    [Fact]
    public void Filter_FiltersElements()
    {
        var input = new[] { 1, 2, 3, 4, 5 };
        var result = StreamProcessing.Filter<int>(n => n > 2, input).ToList();
        Assert.Equal(new[] { 3, 4, 5 }, result);
    }

    [Fact]
    public void Map_TransformsElements()
    {
        var input = new[] { 1, 2, 3 };
        var result = StreamProcessing.Map(n => n * 10, input).ToList();
        Assert.Equal(new[] { 10, 20, 30 }, result);
    }

    [Fact]
    public void Append_ConcatenatesStreams()
    {
        var s1 = new[] { 1, 2 };
        var s2 = new[] { 3, 4 };
        var result = StreamProcessing.Append(s1, s2).ToList();
        Assert.Equal(new[] { 1, 2, 3, 4 }, result);
    }

    // ============================================
    // スライディングウィンドウのテスト
    // ============================================

    [Fact]
    public void Sliding_CreatesWindows()
    {
        var input = new[] { 1, 2, 3, 4, 5 };
        var result = StreamProcessing.Sliding(3, input).Select(w => w.ToList()).ToList();
        Assert.Equal(3, result.Count);
        Assert.Equal(new[] { 1, 2, 3 }, result[0]);
        Assert.Equal(new[] { 2, 3, 4 }, result[1]);
        Assert.Equal(new[] { 3, 4, 5 }, result[2]);
    }

    [Fact]
    public void Pairwise_CreatesPairs()
    {
        var input = new[] { 1, 2, 3, 4 };
        var result = StreamProcessing.Pairwise(input).ToList();
        Assert.Equal(new[] { (1, 2), (2, 3), (3, 4) }, result);
    }

    // ============================================
    // トレンド判定のテスト
    // ============================================

    [Fact]
    public void Trending_DetectsUpwardTrend()
    {
        Assert.True(StreamProcessing.Trending(Seq(0.81m, 0.82m, 0.83m)));
        Assert.False(StreamProcessing.Trending(Seq(0.81m, 0.84m, 0.83m)));
        Assert.False(StreamProcessing.Trending(Seq(0.83m, 0.82m, 0.81m)));
    }

    [Fact]
    public void Declining_DetectsDownwardTrend()
    {
        Assert.True(StreamProcessing.Declining(Seq(0.83m, 0.82m, 0.81m)));
        Assert.False(StreamProcessing.Declining(Seq(0.81m, 0.82m, 0.83m)));
        Assert.False(StreamProcessing.Declining(Seq(0.81m, 0.80m, 0.82m)));
    }

    [Fact]
    public void Stable_DetectsStability()
    {
        Assert.True(StreamProcessing.Stable(Seq(5m, 5m, 5m)));
        Assert.False(StreamProcessing.Stable(Seq(5m, 5m, 6m)));
        Assert.False(StreamProcessing.Stable(Seq(5m, 6m))); // 3つ未満
    }

    // ============================================
    // ストリーム結合のテスト
    // ============================================

    [Fact]
    public void Zip_CombinesStreams()
    {
        var s1 = new[] { 1, 2, 3 };
        var s2 = new[] { "a", "b", "c" };
        var result = StreamProcessing.Zip(s1, s2).ToList();
        Assert.Equal(new[] { (1, "a"), (2, "b"), (3, "c") }, result);
    }

    [Fact]
    public void ZipLeft_TakesLeftValues()
    {
        var s1 = new[] { 1, 2, 3 };
        var s2 = new[] { "a", "b", "c" };
        var result = StreamProcessing.ZipLeft(s1, s2).ToList();
        Assert.Equal(new[] { 1, 2, 3 }, result);
    }

    [Fact]
    public void ZipRight_TakesRightValues()
    {
        var s1 = new[] { 1, 2, 3 };
        var s2 = new[] { "a", "b", "c" };
        var result = StreamProcessing.ZipRight(s1, s2).ToList();
        Assert.Equal(new[] { "a", "b", "c" }, result);
    }

    // ============================================
    // 実用的なストリーム操作のテスト
    // ============================================

    [Fact]
    public void RunningSum_CalculatesCumulativeSum()
    {
        var input = new[] { 1, 2, 3, 4, 5 };
        var result = StreamProcessing.RunningSum(input).ToList();
        Assert.Equal(new[] { 1, 3, 6, 10, 15 }, result);
    }

    [Fact]
    public void MovingAverage_CalculatesAverage()
    {
        var input = new[] { 1m, 2m, 3m, 4m, 5m };
        var result = StreamProcessing.MovingAverage(3, input).ToList();
        Assert.Equal(new[] { 2m, 3m, 4m }, result);
    }

    [Fact]
    public void DistinctConsecutive_RemovesDuplicates()
    {
        var input = new[] { 1, 1, 2, 2, 2, 3, 1, 1 };
        var result = StreamProcessing.DistinctConsecutive(input).ToList();
        Assert.Equal(new[] { 1, 2, 3, 1 }, result);
    }

    [Fact]
    public void Chunk_ChunksStream()
    {
        var input = new[] { 1, 2, 3, 4, 5, 6, 7 };
        var result = StreamProcessing.Chunk(3, input).Select(c => c.ToList()).ToList();
        Assert.Equal(3, result.Count);
        Assert.Equal(new[] { 1, 2, 3 }, result[0]);
        Assert.Equal(new[] { 4, 5, 6 }, result[1]);
        Assert.Equal(new[] { 7 }, result[2]);
    }

    // ============================================
    // 遅延評価のテスト
    // ============================================

    [Fact]
    public void FindFirst_FindsElement()
    {
        var input = new[] { 1, 2, 3, 4, 5 };
        var result = StreamProcessing.FindFirst(n => n > 3, input);
        Assert.Equal(Some(4), result);
    }

    [Fact]
    public void FindFirst_ReturnsNoneWhenNotFound()
    {
        var input = new[] { 1, 2, 3 };
        var result = StreamProcessing.FindFirst(n => n > 10, input);
        Assert.True(result.IsNone);
    }

    [Fact]
    public void TakeWhile_TakesWhileConditionIsTrue()
    {
        var input = new[] { 1, 2, 3, 4, 5 };
        var result = StreamProcessing.TakeWhile<int>(n => n < 4, input).ToList();
        Assert.Equal(new[] { 1, 2, 3 }, result);
    }

    [Fact]
    public void DropWhile_DropsWhileConditionIsTrue()
    {
        var input = new[] { 1, 2, 3, 4, 5 };
        var result = StreamProcessing.DropWhile<int>(n => n < 4, input).ToList();
        Assert.Equal(new[] { 4, 5 }, result);
    }

    // ============================================
    // 通貨交換のテスト
    // ============================================

    [Fact]
    public void Rates_GeneratesExchangeRates()
    {
        var result = StreamProcessing.Rates(StreamProcessing.Currency.USD, StreamProcessing.Currency.EUR)
            .Take(5).ToList();
        Assert.Equal(5, result.Count);
        Assert.True(result.All(r => r > 0m));
    }

    [Fact]
    public void ExchangeIfTrending_TestPureFunction()
    {
        // 純粋関数のテストのみ（F#と同様）
        var trendingRates = Seq(0.81m, 0.82m, 0.83m);
        Assert.True(StreamProcessing.Trending(trendingRates));
    }
}

/// <summary>
/// AsyncStream のテスト
/// </summary>
public class AsyncStreamTests
{
    [Fact]
    public async Task Take_TakesFirstN()
    {
        var stream = RepeatForTest(1);
        var result = await AsyncStream.ToSeqAsync(AsyncStream.Take(3, stream));
        Assert.Equal(3, result.Count);
    }

    [Fact]
    public async Task Filter_FiltersElements()
    {
        var stream = FromEnumerable(new[] { 1, 2, 3, 4, 5 });
        var result = await AsyncStream.ToSeqAsync(AsyncStream.Filter(x => x > 2, stream));
        Assert.Equal(Seq(3, 4, 5), result);
    }

    [Fact]
    public async Task Map_TransformsElements()
    {
        var stream = FromEnumerable(new[] { 1, 2, 3 });
        var result = await AsyncStream.ToSeqAsync(AsyncStream.Map(x => x * 2, stream));
        Assert.Equal(Seq(2, 4, 6), result);
    }

    // ヘルパーメソッド - 有限のストリームを返す
    private static async IAsyncEnumerable<int> RepeatForTest(int value)
    {
        for (var i = 0; i < 10; i++)
        {
            yield return value;
        }
    }

    private static async IAsyncEnumerable<T> FromEnumerable<T>(IEnumerable<T> source)
    {
        foreach (var item in source)
        {
            yield return item;
        }
    }
}

File: app/fsharp/tests/Ch09/StreamProcessingTests.fs
module Ch09.StreamProcessingTests

open Xunit
open Ch09.StreamProcessing

// ============================================
// 基本的な Seq のテスト
// ============================================

[<Fact>]
let ``numbers は 1, 2, 3 を含む`` () =
    let result = numbers |> Seq.toList
    Assert.Equal<int list>([1; 2; 3], result)

[<Fact>]
let ``oddNumbers は奇数のみをフィルタする`` () =
    let input = seq { 1; 2; 3; 4; 5 }
    let result = oddNumbers input |> Seq.toList
    Assert.Equal<int list>([1; 3; 5], result)

[<Fact>]
let ``doubled は各要素を2倍にする`` () =
    let input = seq { 1; 2; 3 }
    let result = doubled input |> Seq.toList
    Assert.Equal<int list>([2; 4; 6], result)

// ============================================
// 無限シーケンスのテスト
// ============================================

[<Fact>]
let ``repeat は要素を無限に繰り返す`` () =
    let result = repeat (seq { 1; 2; 3 }) |> Seq.truncate 8 |> Seq.toList
    Assert.Equal<int list>([1; 2; 3; 1; 2; 3; 1; 2], result)

[<Fact>]
let ``infinite123s は 1, 2, 3 を無限に繰り返す`` () =
    let result = infinite123s |> Seq.truncate 7 |> Seq.toList
    Assert.Equal<int list>([1; 2; 3; 1; 2; 3; 1], result)

[<Fact>]
let ``naturals は自然数の無限シーケンス`` () =
    let result = naturals |> Seq.truncate 5 |> Seq.toList
    Assert.Equal<int list>([1; 2; 3; 4; 5], result)

[<Fact>]
let ``from は指定値から始まる無限シーケンス`` () =
    let result = from 10 |> Seq.truncate 5 |> Seq.toList
    Assert.Equal<int list>([10; 11; 12; 13; 14], result)

// ============================================
// サイコロストリームのテスト
// ============================================

[<Fact>]
let ``infiniteDieCasts は1から6の値を生成する`` () =
    // 新しいシーケンスを毎回作成して10個だけテスト
    let results = infiniteDieCasts |> Seq.truncate 10 |> Seq.toList
    Assert.True(results |> List.forall (fun n -> n >= 1 && n <= 6))

[<Fact>]
let ``castUntil は目標値まで振り続ける`` () =
    // 最大20回に制限
    let results = castUntil 6 |> Seq.truncate 20 |> Seq.toList
    Assert.True(results.Length >= 1 && results.Length <= 20)

// ============================================
// ストリーム操作のテスト
// ============================================

[<Fact>]
let ``take は最初の n 要素を取得する`` () =
    let input = seq { 1; 2; 3; 4; 5 }
    let result = take 3 input |> Seq.toList
    Assert.Equal<int list>([1; 2; 3], result)

[<Fact>]
let ``filter は条件を満たす要素のみを返す`` () =
    let input = seq { 1; 2; 3; 4; 5 }
    let result = filter (fun n -> n > 2) input |> Seq.toList
    Assert.Equal<int list>([3; 4; 5], result)

[<Fact>]
let ``map は各要素を変換する`` () =
    let input = seq { 1; 2; 3 }
    let result = map (fun n -> n * 10) input |> Seq.toList
    Assert.Equal<int list>([10; 20; 30], result)

[<Fact>]
let ``append は2つのストリームを結合する`` () =
    let s1 = seq { 1; 2 }
    let s2 = seq { 3; 4 }
    let result = append s1 s2 |> Seq.toList
    Assert.Equal<int list>([1; 2; 3; 4], result)

// ============================================
// スライディングウィンドウのテスト
// ============================================

[<Fact>]
let ``sliding はスライディングウィンドウを作成する`` () =
    let input = seq { 1; 2; 3; 4; 5 }
    let result = sliding 3 input |> Seq.toList
    Assert.Equal<int list list>([ [1; 2; 3]; [2; 3; 4]; [3; 4; 5] ], result)

[<Fact>]
let ``pairwise は連続する要素のペアを返す`` () =
    let input = seq { 1; 2; 3; 4 }
    let result = pairwise input |> Seq.toList
    Assert.Equal<(int * int) list>([ (1, 2); (2, 3); (3, 4) ], result)

// ============================================
// トレンド判定のテスト
// ============================================

[<Fact>]
let ``trending は上昇トレンドを検出する`` () =
    Assert.True(trending [0.81m; 0.82m; 0.83m])
    Assert.False(trending [0.81m; 0.84m; 0.83m])
    Assert.False(trending [0.83m; 0.82m; 0.81m])

[<Fact>]
let ``declining は下降トレンドを検出する`` () =
    Assert.True(declining [0.83m; 0.82m; 0.81m])
    Assert.False(declining [0.81m; 0.82m; 0.83m])
    Assert.False(declining [0.81m; 0.80m; 0.82m])

[<Fact>]
let ``stable は安定を検出する`` () =
    Assert.True(stable [5m; 5m; 5m])
    Assert.False(stable [5m; 5m; 6m])
    Assert.False(stable [5m; 6m])  // 3つ未満

// ============================================
// ストリーム結合のテスト
// ============================================

[<Fact>]
let ``zip は2つのストリームを結合する`` () =
    let s1 = seq { 1; 2; 3 }
    let s2 = seq { "a"; "b"; "c" }
    let result = zip s1 s2 |> Seq.toList
    Assert.Equal<(int * string) list>([ (1, "a"); (2, "b"); (3, "c") ], result)

[<Fact>]
let ``zipLeft は左側の値を返す`` () =
    let s1 = seq { 1; 2; 3 }
    let s2 = seq { "a"; "b"; "c" }
    let result = zipLeft s1 s2 |> Seq.toList
    Assert.Equal<int list>([1; 2; 3], result)

[<Fact>]
let ``zipRight は右側の値を返す`` () =
    let s1 = seq { 1; 2; 3 }
    let s2 = seq { "a"; "b"; "c" }
    let result = zipRight s1 s2 |> Seq.toList
    Assert.Equal<string list>(["a"; "b"; "c"], result)

// ============================================
// 実用的なストリーム操作のテスト
// ============================================

[<Fact>]
let ``runningSum は累積和を計算する`` () =
    let input = seq { 1; 2; 3; 4; 5 }
    let result = runningSum input |> Seq.toList
    Assert.Equal<int list>([1; 3; 6; 10; 15], result)

[<Fact>]
let ``movingAverage は移動平均を計算する`` () =
    let input = seq { 1m; 2m; 3m; 4m; 5m }
    let result = movingAverage 3 input |> Seq.toList
    // (1+2+3)/3=2, (2+3+4)/3=3, (3+4+5)/3=4
    Assert.Equal<decimal list>([2m; 3m; 4m], result)

[<Fact>]
let ``distinctConsecutive は連続する重複を除去する`` () =
    let input = seq { 1; 1; 2; 2; 2; 3; 1; 1 }
    let result = distinctConsecutive input |> Seq.toList
    Assert.Equal<int list>([1; 2; 3; 1], result)

[<Fact>]
let ``chunk はストリームをチャンク化する`` () =
    let input = seq { 1; 2; 3; 4; 5; 6; 7 }
    let result = chunk 3 input |> Seq.toList
    Assert.Equal<int list list>([ [1; 2; 3]; [4; 5; 6]; [7] ], result)

// ============================================
// 遅延評価のテスト
// ============================================

[<Fact>]
let ``findFirst は最初に条件を満たす要素を見つける`` () =
    let input = seq { 1; 2; 3; 4; 5 }
    let result = findFirst (fun n -> n > 3) input
    Assert.Equal(Some 4, result)

[<Fact>]
let ``findFirst は見つからない場合 None を返す`` () =
    let input = seq { 1; 2; 3 }
    let result = findFirst (fun n -> n > 10) input
    Assert.Equal(None, result)

// ============================================
// 通貨交換のテスト
// ============================================

[<Fact>]
let ``rates は為替レートのストリームを生成する`` () =
    // 5個だけ取得
    let result = rates USD EUR |> Seq.truncate 5 |> Seq.toList
    Assert.Equal(5, List.length result)
    Assert.True(result |> List.forall (fun r -> r > 0m))

[<Fact>]
let ``exchangeIfTrending は上昇トレンドで交換する`` () =
    // 純粋関数のテストのみ
    let trendingRates = [0.81m; 0.82m; 0.83m]
    Assert.True(trending trendingRates)

// ============================================
// AsyncStream のテスト
// ============================================

module AsyncStreamTests =

    open Ch09.AsyncStream

    [<Fact>]
    let ``empty は空のストリームを返す`` () =
        let result = empty<int> |> toListAsync |> Async.RunSynchronously
        Assert.Empty(result)

    [<Fact>]
    let ``singleton は単一要素のストリームを返す`` () =
        let result = singleton 42 |> toListAsync |> Async.RunSynchronously
        Assert.Equal<int list>([42], result)

    [<Fact>]
    let ``fromAsync は Async から AsyncStream を作成する`` () =
        let asyncValue = async { return 123 }
        let result = fromAsync asyncValue |> toListAsync |> Async.RunSynchronously
        Assert.Equal<int list>([123], result)

    [<Fact>]
    let ``take は最初の n 要素を取得する`` () =
        let stream = repeat 1
        let result = stream |> take 3 |> toListAsync |> Async.RunSynchronously
        Assert.Equal<int list>([1; 1; 1], result)
